File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, a: List[int]) -> List[int]:
        #CONCEPT: MONOTINIC DECREASING STACK (CAN BE EQUAL)
        
        res = [0] * len(a)
        
        stk = [(a[0], 0)]
        
        for i, n in enumerate(a):
            while stk and stk[-1][0] < n:
                val, prevIdx = stk.pop()
                res[prevIdx] = i - prevIdx
            stk.append((n, i))
            
        return res
        

File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution {
    public record Entry(int val, int idx) {};
    public int[] dailyTemperatures(int[] t) {
        Stack<Entry> s = new Stack<>();
        int n = t.length;
        int res[] = new int[n];
        for (int i = n-1; i >= 0; i--) {
            // System.out.println(s);
            
            int curr = t[i];
            while (!s.isEmpty() && s.peek().val <= curr){
                s.pop();
            }

            if (s.isEmpty()) res[i] = 0;
            else{
                res[i] = s.peek().idx - i;
            }

            s.push(new Entry(curr, i));
        }

        return res;
    }
}